cpp/src/arrow/compute/exec/hash_join_benchmark.cc

```diff
@@ -0,0 +1,425 @@
+  settings.key_types = key_types;
+
+  HashJoinBasicBenchmarkImpl(st, settings);
+}
+
+template <typename... Args>
+static void BM_HashJoinBasic_Selectivity(benchmark::State& st,
+                                         std::vector<std::shared_ptr<DataType>> key_types,
+                                         Args&&...) {
+  BenchmarkSettings settings;
+  settings.selectivity = static_cast<double>(st.range(0)) / 100.0;
+
+  settings.num_build_batches = static_cast<int>(st.range(1));
+  settings.num_probe_batches = settings.num_build_batches;
+  settings.key_types = key_types;
+
+  HashJoinBasicBenchmarkImpl(st, settings);
+}
+
+template <typename... Args>
+static void BM_HashJoinBasic_JoinType(benchmark::State& st, JoinType join_type,
+                                      Args&&...) {
+  BenchmarkSettings settings;
+  settings.selectivity = static_cast<double>(st.range(0)) / 100.0;
+
+  settings.num_build_batches = static_cast<int>(st.range(1));
+  settings.num_probe_batches = settings.num_build_batches;
+  settings.join_type = join_type;
+
+  HashJoinBasicBenchmarkImpl(st, settings);
+}
+
+static void BM_HashJoinBasic_MatchesPerRow(benchmark::State& st) {
+  BenchmarkSettings settings;
+  settings.cardinality = 1.0 / static_cast<double>(st.range(0));
+
+  settings.num_build_batches = static_cast<int>(st.range(1));
+  settings.num_probe_batches = settings.num_probe_batches;
+
+  HashJoinBasicBenchmarkImpl(st, settings);
+}
+
+static void BM_HashJoinBasic_PayloadSize(benchmark::State& st) {
+  BenchmarkSettings settings;
+  int32_t payload_size = static_cast<int32_t>(st.range(0));
+  settings.probe_payload_types = {fixed_size_binary(payload_size)};
+  settings.cardinality = 1.0 / static_cast<double>(st.range(1));
+
+  settings.num_build_batches = static_cast<int>(st.range(2));
+  settings.num_probe_batches = settings.num_probe_batches;
+
+  HashJoinBasicBenchmarkImpl(st, settings);
+}
+
+static void BM_HashJoinBasic_ProbeParallelism(benchmark::State& st) {
+  BenchmarkSettings settings;
+  settings.num_threads = static_cast<int>(st.range(0));
+  settings.num_build_batches = static_cast<int>(st.range(1));
+  settings.num_probe_batches = settings.num_build_batches * 8;
+
+  HashJoinBasicBenchmarkImpl(st, settings);
+}
+
+static void BM_HashJoinBasic_BuildParallelism(benchmark::State& st) {
+  BenchmarkSettings settings;
+  settings.num_threads = static_cast<int>(st.range(0));
+  settings.num_build_batches = static_cast<int>(st.range(1));
+  settings.num_probe_batches = settings.num_threads;
+
+  HashJoinBasicBenchmarkImpl(st, settings);
+}
+
+static void BM_HashJoinBasic_NullPercentage(benchmark::State& st) {
+  BenchmarkSettings settings;
+  settings.null_percentage = static_cast<double>(st.range(0)) / 100.0;
+
+  HashJoinBasicBenchmarkImpl(st, settings);
+}
+
+std::vector<std::string> keytypes_argnames = {"Hashtable krows"};
+BENCHMARK_CAPTURE(BM_HashJoinBasic_KeyTypes, "{int32}", {int32()})
```

Comment:
Well I wanted it to be consistent with the multikey ones. Should I drop the braces on those? Or are braces ok with those and keep single-key tests without them?

Reply:
I don't feel terribly strongly about this so if you want to keep the braces I'd be ok.  However, I do agree it should be consistent.  So either braces on both or neither.  At the time I was thinking neither.  In my mind I was thinking about automated tools parsing the benchmark results but in retrospect I suppose most of those tools will be consuming the JSON so it is a moot point.

Reply:
Ok, I think I'll leave them just because I think it makes the legend clearer on the graph that that's the key type. I did make the python script remove the quoting from the json.